pkg/container/versioned/value.go

```diff
@@ -0,0 +1,440 @@
+	// 'version' is the version this VersionHold keeps from being released
+	version version
+	// Coordinator of this VersionHold, if any. Used for closing if non-nil.
+	// Atomic due to nilling and for copy prevention.
+	coordinator atomic.Pointer[Coordinator]
+}
+
+func (h *VersionHold) String() string {
+	if h == nil {
+		return "version: <nil>"
+	}
+	return fmt.Sprintf("version: %d", h.version)
+}
+
+// versionHoldFinalizer is used to warn about missing Close() calls.
+func versionHoldFinalizer(h *VersionHold) {
+	slog.Error("Hold for SelectorCache version not closed.", logfields.Version, h.version, logfields.LogSubsys, "containers/versioned")
+	h.Close()
+}
+
+func newVersionHold(version version, coordinator *Coordinator) *VersionHold {
+	// hold on maxVersion never expires
+	if version == maxVersion {
+		coordinator = nil
+	}
+	h := &VersionHold{version: version}
+	h.coordinator.Store(coordinator)
+	if coordinator != nil {
+		// Set a finalizer to catch unclosed holds. The finalizer function complains
+		// loudly, so that we do not rely the finalizer for closing.
+		runtime.SetFinalizer(h, versionHoldFinalizer)
+	}
+	return h
+}
+
+// Latest returns a VersionHold for the latest version of current/non-removed values
+// Only to be used in cases where the latest values are needed and no transactionality is required.
+func Latest() *VersionHold {
+	return newVersionHold(maxVersion, nil)
+}
+
+func (h *VersionHold) IsValid() bool {
+	return h != nil && (h.version == maxVersion || h.coordinator.Load() != nil)
+}
+
+// Close releases the held version for removal once no hold for it are no longer held.
+// This may not be called while holding any locks that the 'closer' function passed to the
+// coordinator may take!
+func (h *VersionHold) Close() (err error) {
+	if !h.IsValid() {
+		return errInvalidVersion
+	}
+	coordinator := h.coordinator.Load()
+	if coordinator != nil {
+		runtime.SetFinalizer(h, nil)
+		err = coordinator.releaseVersion(h.version)
+		h.coordinator.Store(nil)
+	}
+	return err
+}
+
+type atomicVersion struct {
+	version atomic.Uint64
+}
+
+func (a *atomicVersion) load() version {
+	return version(a.version.Load())
+}
+
+func (a *atomicVersion) store(version version) {
+	a.version.Store(uint64(version))
+}
+
+func (a *atomicVersion) compareAndSwap(old, new version) bool {
+	return a.version.CompareAndSwap(uint64(old), uint64(new))
+}
+
+type versionCount struct {
+	version version
+	count   int
+}
+
+// Coordinator defines a common version number space for multiple versioned.Values,
+// and provides facilities for cleaning out stale versions.
+// The Value's are not directly managed by the Coordinator, but all the values under coordination
+// should be cleaned by the 'cleaner' function given to the Coordinator.
+type Coordinator struct {
+	// version is the version number of the last applied change
+	version atomicVersion
+
+	// cleaner is called with the earliest version that must be kept
+	cleaner func(KeepVersion)
```

Comment:
add a note that the cleaner may be called concurrently, which the func needs to handle itself

Reply:
Added a comment.